pkg/controller/workload/mpijob/mpijob_controller.go

```diff
@@ -0,0 +1,579 @@
+		client.MatchingFields{ownerKey: job.Name}); err != nil {
+		log.Error(err, "Unable to list child workloads")
+		return nil, err
+	}
+
+	for i := range workloads.Items {
+		w := &workloads.Items[i]
+		owner := metav1.GetControllerOf(w)
+		// Indexes don't work in unit tests, so we explicitly check for the
+		// owner here.
+		if owner.Name != job.Name {
+			continue
+		}
+		if match == nil && jobAndWorkloadEqual(job, w) {
+			match = w
+		} else {
+			toDelete = append(toDelete, w)
+		}
+	}
+
+	// If there is no matching workload and the job is running, suspend it.
+	if match == nil && !jobSuspended(job) {
+		log.V(2).Info("job with no matching workload, suspending")
+		var w *kueue.Workload
+		if len(workloads.Items) == 1 {
+			// The job may have been modified and hence the existing workload
+			// doesn't match the job anymore. All bets are off if there are more
+			// than one workload...
+			w = &workloads.Items[0]
+		}
+		if err := r.stopJob(ctx, w, job, "No matching Workload"); err != nil {
+			log.Error(err, "stopping job")
+		}
+	}
+
+	// Delete duplicate workload instances.
+	existedWls := 0
+	for i := range toDelete {
+		err := r.client.Delete(ctx, toDelete[i])
+		if err == nil || !apierrors.IsNotFound(err) {
+			existedWls++
+		}
+		if err != nil && !apierrors.IsNotFound(err) {
+			log.Error(err, "Failed to delete workload")
+		}
+		if err == nil {
+			r.record.Eventf(job, corev1.EventTypeNormal, "DeletedWorkload",
+				"Deleted not matching Workload: %v", workload.Key(toDelete[i]))
+		}
+	}
+
+	if existedWls != 0 {
+		if match == nil {
+			return nil, fmt.Errorf("no matching workload was found, tried deleting %d existing workload(s)", existedWls)
+		}
+		return nil, fmt.Errorf("only one workload should exist, found %d", len(workloads.Items))
+	}
+
+	return match, nil
+}
+
+func ConstructWorkloadFor(ctx context.Context, client client.Client,
+	job *kubeflow.MPIJob, scheme *runtime.Scheme) (*kueue.Workload, error) {
+	w := &kueue.Workload{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      GetWorkloadNameForMPIJob(job.Name),
+			Namespace: job.Namespace,
+		},
+		Spec: kueue.WorkloadSpec{
+			QueueName: queueName(job),
+		},
+	}
+
+	highestPriorityClassName := ""
+	highestPriority := int32(0)
+
+	for _, mpiReplicaType := range orderedReplicaTypes(&job.Spec) {
+		podSet := kueue.PodSet{
+			Name:  strings.ToLower(string(mpiReplicaType)),
+			Spec:  *job.Spec.MPIReplicaSpecs[mpiReplicaType].Template.Spec.DeepCopy(),
+			Count: podsCount(&job.Spec, mpiReplicaType),
+		}
+		w.Spec.PodSets = append(w.Spec.PodSets, podSet)
+
+		priorityClassName, p, err := utilpriority.GetPriorityFromPriorityClass(
+			ctx, client, job.Spec.MPIReplicaSpecs[mpiReplicaType].Template.Spec.PriorityClassName)
+		if err != nil {
+			return nil, err
+		}
+		if p > highestPriority {
+			highestPriority = p
+			highestPriorityClassName = priorityClassName
+		}
+	}
+	// Populate priority from priority class.
+	w.Spec.Priority = &highestPriority
```

Comment:
There is another field that has priority: `.spec.runPolicy.schedulingPolicy`. That should take precedence.

Reply:
I only do lookup for the field now which seems the dedicated mechanism. Let me know if it requires some fallback at all.

Reply:
The mpi-operator fallbacks to replica priorities. So I prefer the same behavior in kueue as well.

https://github.com/kubeflow/mpi-operator/blob/5946ef4157599a474ab82ff80e780d5c2546c9ee/pkg/controller/podgroup.go#L69-L86

WDYT? @alculquicondor @mimowo

Reply:
+1
I guess the precedence is: runpolicy, launcher, worker; regardless of the priority values.

Reply:
Done + unit test.